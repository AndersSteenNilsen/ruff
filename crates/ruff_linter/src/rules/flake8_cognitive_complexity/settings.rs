//! Settings for the `flake8_cognitive_complexity` plugin.

use crate::display_settings;
use ruff_macros::CacheKey;
use std::fmt::{Display, Formatter};

#[derive(Debug, CacheKey)]
pub struct Settings {
    pub max_cognitive_complexity: usize,
}

pub const DEFAULT_MAX_COGNITIVE_COMPLEXITY: usize = 7;

impl Default for Settings {
    fn default() -> Self {
        Self {
            max_cognitive_complexity: DEFAULT_MAX_COGNITIVE_COMPLEXITY,
        }
    }
}

impl Display for Settings {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        display_settings! {
            formatter = f,
            namespace = "linter.flake8_cognitive_complexity",
            fields = [
                self.max_cognitive_complexity
            ]
        }
        Ok(())
    }
}
