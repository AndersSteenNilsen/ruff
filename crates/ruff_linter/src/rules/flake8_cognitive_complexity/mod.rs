//! Rules from [flake8-cognitive-complexity](https://pypi.org/project/flake8-cognitive-complexity/).
pub(crate) mod rules;
pub mod settings;

#[cfg(test)]
mod tests {
    use std::path::Path;

    use anyhow::Result;
    use test_case::test_case;

    use crate::assert_messages;
    use crate::registry::Rule;
    use crate::settings::LinterSettings;
    use crate::test::test_path;

    #[test_case(0)]
    #[test_case(3)]
    #[test_case(10)]
    fn max_cognitive_complexity_zero(max_cognitive_complexity: usize) -> Result<()> {
        let snapshot = format!("max_cognitive_complexity_{max_cognitive_complexity}");
        let diagnostics = test_path(
            Path::new("flake8_cognitive_complexity/CCR001.py"),
            &LinterSettings {
                flake8_cognitive_complexity: super::settings::Settings {
                    max_cognitive_complexity,
                },
                ..LinterSettings::for_rules(vec![Rule::CognitiveComplexStructure])
            },
        )?;
        assert_messages!(snapshot, diagnostics);
        Ok(())
    }
}
