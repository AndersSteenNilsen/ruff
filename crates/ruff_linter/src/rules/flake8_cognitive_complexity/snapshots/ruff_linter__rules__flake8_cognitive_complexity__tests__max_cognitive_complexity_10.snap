---
source: crates/ruff_linter/src/rules/flake8_cognitive_complexity/mod.rs
---

